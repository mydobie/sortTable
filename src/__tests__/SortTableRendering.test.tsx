/* eslint-disable react/react-in-jsx-scope */

import { axe } from 'jest-axe';
import { sortTable, data, headers } from './helpers/helpers';

describe('Sort Table Rendering', () => {
  test('Snapshot test', () => {
    expect(
      sortTable({ showPagination: true, showFilter: true })
    ).toMatchSnapshot();
  });

  test('Expected number of columns and cells', () => {
    const wrapper = sortTable();
    expect(wrapper.find('tbody tr')).toHaveLength(data.length);
    expect(wrapper.find('thead tr').children()).toHaveLength(headers.length);
  });

  test('Header cells are marked with th tag', () => {
    expect(sortTable().find('thead th')).toHaveLength(headers.length);
  });

  test('Row headers are marked with th tags', () => {
    // Test set-up - ensure that there is only one row header
    const firstRow = sortTable().find('tbody tr').first();

    expect(firstRow.find('th')).toHaveLength(1);
    expect(firstRow.find('td')).toHaveLength(headers.length - 1);

    // Test Start
    const rowHeaderIndex = headers.findIndex((row) => row.rowheader);
    expect(firstRow.children().at(rowHeaderIndex).type()).toEqual('th');
  });

  test('Components are rendered in a data cell', () => {
    const dataHtml = [
      { id: 1, name: <strong>Cheese</strong>, price: '$4.90', stock: 20 },
    ];
    const wrapper = sortTable({
      tableData: dataHtml,
    });
    const nameIndex = headers.findIndex((row) => row.key === 'name');
    expect(nameIndex).not.toBeUndefined();

    expect(
      wrapper.find('tbody tr').children().at(nameIndex).find('strong')
    ).toHaveLength(1);
  });

  test('DangerouslySetInnerHTML allows for rendered html', () => {
    const dataHtml = [
      { id: 1, name: '<strong>Cheese</strong>', price: '$4.90', stock: 20 },
    ];
    const wrapper = sortTable({
      dangerouslySetInnerHTML: true,
      tableData: dataHtml,
    });
    const nameIndex = headers.findIndex((row) => row.key === 'name');
    expect(nameIndex).not.toBeUndefined();

    expect(wrapper.find('tbody tr').children().at(nameIndex).html()).toMatch(
      /<span><strong>Cheese<\/strong><\/span>/
    );
  });

  test('Filtering text box is not shown by default', () => {
    expect(sortTable().find('[data-filter]')).toHaveLength(0);
  });

  test('Pagination is not shown by default', () => {
    expect(sortTable().find('[data-pagination]')).toHaveLength(0);
    expect(sortTable().find('[data-sort-number-of-inputs]')).toHaveLength(0);
  });

  test('Table caption is shown when value is passed', () => {
    const myCaption = 'my caption';
    const wrapper = sortTable({ caption: myCaption });
    expect(wrapper.find('caption')).toHaveLength(1);
    expect(wrapper.find('caption').text()).toEqual(myCaption);
  });

  test('Empty caption tags are not shown', () => {
    expect(sortTable().find('caption')).toHaveLength(0);
  });

  test('Is Loading default message is shown', () => {
    // Test set-up - ensure loading isn't shown by default
    expect(sortTable().find('Loading')).toHaveLength(0);

    // Test Start
    expect(sortTable({ isLoading: true }).find('Loading')).toHaveLength(1);
    // verify data is no shown
    expect(
      sortTable({ isLoading: true }).find('[data-sorttable-data-cell]')
    ).toHaveLength(0);
  });

  test('Is Loading custom message is shown', () => {
    const customLoad = <div id='MyCustomLoading'>My loader</div>;
    expect(
      sortTable({ isLoading: true, isLoadingMessage: customLoad }).find(
        '#MyCustomLoading'
      )
    ).toHaveLength(1);
  });

  describe('Custom CSS', () => {
    test('Table header class name is set when passed', () => {
      const wrapper = sortTable({ headerClassName: 'myCustomHeaderClass' });
      expect(wrapper.find('thead.myCustomHeaderClass')).toHaveLength(1);
    });

    test('Responsive data attribute is set', () => {
      // verify responsive attribute is not set by default
      expect(sortTable().find('table[data-sort-responsive]')).toHaveLength(0);

      // applied with prop
      expect(
        sortTable({ isResponsive: true }).find('table[data-sort-responsive]')
      ).toHaveLength(1);
    });

    test('Custom table class name is set', () => {
      expect(
        sortTable({ tableClassName: 'myCustomClass' }).find(
          'table.myCustomClass'
        )
      ).toHaveLength(1);
    });

    test('Custom class name is added to column header', () => {
      // Test set-up - ensure a header className is set
      const index = headers.findIndex((header) => header.className);
      const { className } = headers[index];
      expect(className).not.toBeUndefined();

      // Test start
      expect(sortTable().find(`th.${className}`)).toHaveLength(1);
    });

    test('Custom style is added to column header', () => {
      // Test set-up - ensure a header className is set
      const index = headers.findIndex((header) => header.style);
      const color = headers[index].style?.color;
      expect(color).not.toBeUndefined();

      // Test start
      expect(sortTable().find('th').at(index).prop('style')).toHaveProperty(
        'color',
        color
      );
    });

    test('Non empty cells does not have custom empty css class name', () => {
      const urlIndex = headers.findIndex((header) => header.key === 'url');
      const index = data.findIndex(
        (row) => row.url !== undefined && row.url !== ''
      );

      expect(urlIndex).not.toEqual(-1);
      expect(index).not.toEqual(-1);

      const wrapper = sortTable({ emptyCellClassName: 'myCustomCSSClassName' });

      expect(
        wrapper.find('tbody tr').at(index).childAt(urlIndex).prop('className')
      ).not.toEqual('myCustomCSSClassName');
    });

    test('Empty cell class is applied to undefined value', () => {
      const urlIndex = headers.findIndex((header) => header.key === 'url');
      const index = data.findIndex((row) => row.url === undefined);

      expect(urlIndex).not.toEqual(-1);
      expect(index).not.toEqual(-1);

      const wrapper = sortTable({ emptyCellClassName: 'myCustomCSSClassName' });

      expect(
        wrapper.find('tbody tr').at(index).childAt(urlIndex).prop('className')
      ).toEqual('myCustomCSSClassName');
    });

    test('Empty cell class is applied to value that is empty', () => {
      const urlIndex = headers.findIndex((header) => header.key === 'url');
      const index = data.findIndex((row) => row.url === '');

      expect(urlIndex).not.toEqual(-1);
      expect(index).not.toEqual(-1);

      const wrapper = sortTable({
        emptyCellClassName: 'myCustomCSSClassName',
      });

      expect(
        wrapper.find('tbody tr').at(index).childAt(urlIndex).prop('className')
      ).toEqual('myCustomCSSClassName');
    });
  });

  describe('Accessibility', () => {
    test('Is accessible with default settings', async () => {
      const results = await axe(`<main>${sortTable()}</main>`);
      expect(results).toHaveNoViolations();
    });

    test('Is accessible with filtering', async () => {
      const results = await axe(
        `<main>${sortTable({ showFilter: true })}</main>`
      );
      expect(results).toHaveNoViolations();
    });

    test('Is accessible with pagination buttons', async () => {
      const results = await axe(
        `<main>${sortTable({ showPagination: true })}</main>`
      );
      expect(results).toHaveNoViolations();
    });

    test('Is accessible with pagination drop down', async () => {
      const wrapper = sortTable({ showPagination: true }, { viewSet: 1 });
      const results = await axe(`<main>${wrapper.html()}</main>`);
      expect(wrapper.find('[data-pagination-select]')).toHaveLength(1);

      expect(results).toHaveNoViolations();
    });
  });
});
