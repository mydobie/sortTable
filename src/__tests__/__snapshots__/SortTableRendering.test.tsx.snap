// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort Table Rendering Snapshot test 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <div
      class="row"
      style="margin-bottom: 15px;"
    >
      <div
        class="col-sm"
      >
        <div
          data-sort-number-of-inputs="true"
        >
          Show
          <select
            aria-label="Number of items shown"
            class="form-control form-select"
            style="width: 75px; display: inline-block; margin-left: 0.75em; margin-right: 0.75em;"
          >
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="4"
            >
              4
            </option>
            <option
              value="10"
            >
              10
            </option>
            <option
              value=""
            >
              All
            </option>
          </select>
          results
        </div>
      </div>
      <div
        class="col-sm"
        style="text-align: right;"
      >
        <div
          class="form-inline justify-content-end"
          data-filter="true"
        >
          <div
            class="form-group"
          >
            <label
              for="sortTableFilterInput"
            >
              Filter
            </label>
            <input
              class="form-control"
              data-filter-input="true"
              id="sortTableFilterInput"
              style="margin-left: 0.75em; display: inline-block; width: 250px;"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="row"
      style=""
    >
      <table
        aria-describedby="sortTableRowsShownSummary"
        aria-rowcount="12"
        class="table undefined"
        id="sortTable"
      >
        <thead>
          <tr
            aria-rowindex="1"
          >
            <th
              class="myCustomPriceClass "
              scope="col"
            >
              <button
                style="background: none; width: 100%; text-align: left;"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  data-icontype="sortable"
                  style="color: rgb(204, 204, 204); display: inline-block; padding-right: 5px; transform: scale(1);"
                >
                  arrow-down-up.svg
                </svg>
                Price
              </button>
            </th>
            <th
              class="undefined "
              scope="col"
            >
              <button
                style="background: none; width: 100%; text-align: left;"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  data-icontype="sortable"
                  style="color: rgb(204, 204, 204); display: inline-block; padding-right: 5px; transform: scale(1);"
                >
                  arrow-down-up.svg
                </svg>
                Product Name
              </button>
            </th>
            <th
              class="undefined "
              scope="col"
            >
              <button
                style="background: none; width: 100%; text-align: left;"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  data-icontype="sortable"
                  style="color: rgb(204, 204, 204); display: inline-block; padding-right: 5px; transform: scale(1);"
                >
                  arrow-down-up.svg
                </svg>
                Stock
              </button>
            </th>
            <th
              class="undefined "
              scope="col"
            >
              <button
                style="background: none; width: 100%; text-align: left;"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  data-icontype="sortable"
                  style="color: rgb(204, 204, 204); display: inline-block; padding-right: 5px; transform: scale(1);"
                >
                  arrow-down-up.svg
                </svg>
                Sale day
              </button>
            </th>
            <th
              class="undefined "
              scope="col"
              style="color: purple;"
            >
              Link
            </th>
          </tr>
        </thead>
        <tbody
          data-testid="sortTableBody"
        >
          <tr>
            <td
              class=" "
              data-sorttable-data-cell="true"
            >
              $4.90
            </td>
            <th
              class=" "
              data-sorttable-data-cell="true"
              scope="row"
            >
              Cheese
            </th>
            <td
              class=" "
              data-sorttable-data-cell="true"
            >
              20
            </td>
            <td
              class=" "
              data-sorttable-data-cell="true"
            >
              Friday
            </td>
            <td
              class="undefined "
              data-sorttable-data-cell="true"
            >
              
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="row"
      style=""
    >
      <div
        class="col-sm"
        data-pagination-summary="true"
        id="sortTableRowsShownSummary"
      >
        <span>
          1 - 1 of 11 entries.
        </span>
      </div>
      <div
        class="col-sm"
      >
        <div>
          <div
            class="text-right text-end"
            data-skip-pagination-wrapper="true"
          >
            <a
              data-skip-pagination-link="true"
              href="#sortTableEndPagination"
            >
              Skip pagination
            </a>
          </div>
          <nav
            aria-label="Table pagination"
          >
            <ul
              class="pagination pagination-sm justify-content-end"
              data-pagination="true"
            >
              <li
                class="page-item disabled "
                data-pagination-previous-button="true"
              >
                <button
                  aria-label="Previous page"
                  class="page-link "
                  disabled=""
                  type="button"
                >
                  <span
                    aria-hidden="true"
                  >
                    «
                  </span>
                </button>
              </li>
              <li>
                <select
                  aria-label="Show page number"
                  class="form-select form-select-sm custom-select custom-select-sm"
                  data-pagination-select="true"
                  style="width: 6em;"
                >
                  <option
                    value="1"
                  >
                    1
                  </option>
                  <option
                    value="2"
                  >
                    2
                  </option>
                  <option
                    value="3"
                  >
                    3
                  </option>
                  <option
                    value="4"
                  >
                    4
                  </option>
                  <option
                    value="5"
                  >
                    5
                  </option>
                  <option
                    value="6"
                  >
                    6
                  </option>
                  <option
                    value="7"
                  >
                    7
                  </option>
                  <option
                    value="8"
                  >
                    8
                  </option>
                  <option
                    value="9"
                  >
                    9
                  </option>
                  <option
                    value="10"
                  >
                    10
                  </option>
                  <option
                    value="11"
                  >
                    11
                  </option>
                </select>
              </li>
              <li
                class="page-item  "
                data-pagination-next-button="true"
              >
                <button
                  aria-label="Next page"
                  class="page-link "
                  type="button"
                >
                  <span
                    aria-hidden="true"
                  >
                    »
                  </span>
                </button>
              </li>
            </ul>
          </nav>
          <div
            id="sortTableEndPagination"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
